spec/index.adoc

```diff
@@ -1674,6 +1674,96 @@ Any user can top-up any canister this way.
 
 This method is only available in local development instances.
 
+
+[#ic-bitcoin-canisters]
+== The IC bitcoin canisters
+
+Similar to the <<ic-management-canister,management canister>>, the Bitcoin functionality is also exposed through a facade, which we refer to as the "Bitcoin canisters".
+There are two Bitcoin canisters differing only in that one operates on the real Bitcoin blockchain whereas the other only works with the Bitcoin testnet.
+The Bitcoin canisters can be accessed using their addresses:
+
+- The canister address of the Bitcoin canister that uses the real Bitcoin blockchain is `TBD`.
+- The canister address of the Bitcoin canister that uses the Bitcoin testnet is `TBD`.
+
+It is *not* possible to interact with either Bitcoin canister via external requests (i.e., ingress messages).
```

Comment:
Also worth mentioning that it is only possible to interact with this canister on specific subnets (i.e. bitcoin-enabled subnets).

This might potentially be a can of worms, as now we'll need to define what a bitcoin-enabled subnet is, and also we need to introduce the concept of subnets, which this specification has been trying to avoid for the longest time.

Reply:
Hhmm... so what should we do here? It's impossible to talk about this without talking about subnets.

Reply:
I don't think there's a way around not talking about subnets either. I'd be curious what Bjoern thinks here.

Reply:
I think that is breaking consistency of the interface. If the interface should not be available everywhere, in particular via ingress, then we should maybe do this as an extension to the System API instead of as a canister facade?

An alternative would be to specify a `NotAllowed` error type for the return value. The canister facade is free to return that value, but we should not have this fail as "not addressable" from some sources.

Reply:
Can the canister methods inspect the caller to see if the message is ingress or not? That's a question, not a suggestion.

Reply:
I think ingress message has to go through boundary node, so we can simply disable routing in the boundary nodes. Similar to the management canister, without an effective canister id, the ingress message cannot call the management canister either.

Reply:
> I think that is breaking consistency of the interface. If the interface should not be available everywhere, in particular via ingress, then we should maybe do this as an extension to the System API instead of as a canister facade?

We already do this for a number of methods on the management canisters, like `raw_rand`., but I'd add that not supporting ingress was more out of precaution - there's no fundamental reason why we can't/shouldn't expose it to preserve the symmetry as far as I know.

@THLO aren't we going to be charging cycles for these calls?

Reply:
Yes, the plan is to charge cycles for these calls (which obviously won't work for ingress messages).

Reply:
But if we anyway plan to charge cycles, then don't we anyway need an error code `InsufficientFunds` (or similar) and we can just reject ingress messages consistently?

Reply:
I agree. There is one caveat here that's worth mentioning: if we introduce an application-level error to indicate that cycles are insufficient, then we have two choices when it comes to ingress messages:

- Accept ingress messages in `canister_inspect_message`, and then reject them with the proper application-level error. This preserves the consistency between ingress and canister calls at the expense of not being able to filter out the ingress message early.
- Reject ingress messages in `canister_inspect_message`, which would be better performance-wise, but then the behavior of ingress messages (receiving a reject) and canisters (receiving an application-level error) would be inconsistent.

I wish candid interfaces had a way of specifying rejects, along with their structure. That way, we could reject with a candid error and clients would know from the interface to expect that, but that's beyond the scope of this PR.

Reply:
Personally, I would go for consistency - at the risk of paying a (hopefully small) performance penalty.

Reply:
I don't think there is a choice:

Rejecting intrercanister messages without cycles in the message handler is only reasonable because the sending canister paid an upfront cost for the message, so this is not an efficient cycle drain attack.

(I expect that eventually we want to refine this and make the first _n_ instructions of an inter-canister message handler free for the canister and implicitly covered by the cost of sending the message, enough for a typical access control check, so that an attackers cannot even run an inefficient cycle drain attack, but that is just an aside.)

But for ingress messages, no cycles were paid, and the receiving canister would have to pay for it, and that's _precisely_ the reason we have introduced the `inspect_message` hook - else the cost of draining cycles is just the cost of HTTP (and maybe the rate limits).

It may be slightly inconsistent (nobody says that `inspect_message` is great), but it _is_ the best practice for canisters, and I don't think we should do it differently for “system-provided” canisters, as that would be another inconsistency.